src/easylink/implementation.py

```diff
@@ -57,46 +91,67 @@ def validate(self) -> list[str]:
     ##################
 
     def _load_metadata(self) -> dict[str, str]:
+        """Loads the metadata for this Implementation instance."""
         metadata = load_yaml(paths.IMPLEMENTATION_METADATA)
         return metadata[self.name]
 
     def _validate_expected_step(self, logs: list[str]) -> list[str]:
+        """Validates that the Implementation is responsible for the correct steps."""
         if not set(self.schema_steps) == set(self.metadata_steps):
             logs.append(
                 f"Pipeline configuration nodes {self.schema_steps} do not match metadata steps {self.metadata_steps}."
             )
         return logs
 
     def _validate_container_exists(self, logs: list[str]) -> list[str]:
+        """Validates that the container for this Implementation exists."""
         err_str = f"Container '{self.singularity_image_path}' does not exist."
         if not Path(self.singularity_image_path).exists():
             logs.append(err_str)
         return logs
 
     def _get_env_vars(self, implementation_config: LayeredConfigTree) -> dict[str, str]:
+        """Gets the environment variables relevant to this Implementation."""
         env_vars = self._metadata.get("env", {})
         env_vars.update(implementation_config.get("configuration", {}))
         return env_vars
 
     @property
     def singularity_image_path(self) -> str:
+        """The path to the Singularity image for this Implementation."""
         return self._metadata["image_path"]
 
     @property
     def script_cmd(self) -> str:
+        """The command to run inside of the container for this Implementation."""
         return self._metadata["script_cmd"]
 
     @property
     def outputs(self) -> dict[str, list[str]]:
+        """The outputs expected from this Implementation."""
         return self._metadata["outputs"]
 
 
 class NullImplementation:
-    """A NullImplementation is used to represent a step that does not have an implementation.
+    """A representation of a :class:`~easylink.step.Step` that does not have an :class:`Implementation`.
```

Comment:
```suggestion
    """An object with some of the interface of :class:`Implementation` that represents that no container needs to run.
    
    The primary use case for this class is when adding an :class:`~easylink.step.IOStep` -
    which does not have a corresponding :class:`Implementation` - to an
    :class:`~easylink.graph_components.ImplementationGraph`. Adding a new node
    requires the ``implementation`` attribute to be an object with input and output slot names.
```

Slight edit and moved the notes section up, since I think it's very important to understanding why this class exists.